src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java

```diff
@@ -360,16 +360,19 @@ protected GitTool resolveGitTool(String gitTool) {
      */
     @CheckForNull
     @Override
-    protected SCMRevision retrieve(@NonNull final SCMHead head, @NonNull TaskListener listener)
+    protected SCMRevision retrieve(@NonNull final SCMHead head, @NonNull final TaskListener listener)
             throws IOException, InterruptedException {
-        final GitSCMSourceContext context = new GitSCMSourceContext<>(null, SCMHeadObserver.none()).withTraits(getTraits());
+        GitSCMSourceContext context = new GitSCMSourceContext<>(null, SCMHeadObserver.none()).withTraits(getTraits());
         GitSCMTelescope telescope = GitSCMTelescope.of(this);
         if (telescope != null) {
             String remote = getRemote();
             StandardUsernameCredentials credentials = getCredentials();
             telescope.validate(remote, credentials);
             return telescope.getRevision(remote, credentials, head);
         }
+        if (head instanceof GitSCMHeadMixin) {
+            context = context.withoutRefSpecs().withRefSpec(((GitSCMHeadMixin) head).getRef());
```

Comment:
What about somebody who has added additional ref specs in a trait... will those refspecs be removed now?

Reply:
Yes I was a bit hesitant on that myself, `context.withoutRefSpecs()` does clear all refspecs from the traits.
But from what I can tell, in this context you are after a specific `SCMHead` so wouldn't all other refspecs just be a waste of time and bandwidth?

Reply:
A comment explaining the logic as to why it is ok to strip it out here would be ideal then